flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetRecordFormat.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.parquet.avro.AvroParquetReader;
+import org.apache.parquet.hadoop.ParquetReader;
+import org.apache.parquet.io.DelegatingSeekableInputStream;
+import org.apache.parquet.io.InputFile;
+import org.apache.parquet.io.SeekableInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** */
+public class AvroParquetRecordFormat<E> implements StreamFormat<E> {
+
+    private static final long serialVersionUID = 1L;
+
+    static final Logger LOG = LoggerFactory.getLogger(AvroParquetRecordFormat.class);
+
+    private final TypeInformation<E> type;
+
+    AvroParquetRecordFormat(TypeInformation<E> type) {
+        this.type = type;
+    }
+
+    /**
+     * Creates a new reader to read avro {@link GenericRecord} from Parquet input stream.
+     *
+     * <p>Several wrapper classes haven be created to Flink abstraction become compatible with the
+     * parquet abstraction. Please refer to the inner classes {@link AvroParquetRecordReader},
+     * {@link ParquetInputFile}, {@link FSDataInputStreamAdapter} for details.
+     */
+    @Override
+    public Reader<E> createReader(
+            Configuration config, FSDataInputStream stream, long fileLen, long splitEnd)
+            throws IOException {
+
+        // current version does not support splitting.
+        checkNotSplit(fileLen, splitEnd);
+
+        return new AvroParquetRecordReader<E>(
+                AvroParquetReader.<E>builder(new ParquetInputFile(stream, fileLen))
+                        .withDataModel(getDataModel())
+                        .build());
+    }
+
+    /**
+     * Restores the reader from a checkpointed position. It is in fact identical since only {@link
+     * CheckpointedPosition#NO_OFFSET} as the {@code restoredOffset} is support.
+     */
+    @Override
+    public Reader<E> restoreReader(
+            Configuration config,
+            FSDataInputStream stream,
+            long restoredOffset,
+            long fileLen,
+            long splitEnd)
+            throws IOException {
+
+        // current version does not support splitting.
+        checkNotSplit(fileLen, splitEnd);
+
+        checkArgument(
+                restoredOffset == CheckpointedPosition.NO_OFFSET,
+                "The restoredOffset should always be NO_OFFSET");
+
+        return createReader(config, stream, fileLen, splitEnd);
+    }
+
+    @VisibleForTesting
+    GenericData getDataModel() {
+        Class<E> typeClass = getProducedType().getTypeClass();
+        if (org.apache.avro.specific.SpecificRecordBase.class.isAssignableFrom(typeClass)) {
+            return SpecificData.get();
+        } else if (org.apache.avro.generic.GenericRecord.class.isAssignableFrom(typeClass)) {
+            return GenericData.get();
+        } else {
+            return ReflectData.get();
+        }
+    }
```

Comment:
We should inject the model with the factory methods. There we already have the knowledge where it's a specific, generic, or reflective type and we don't need to infer that here.

Reply:
This is the question of API design: Do we want to make the internal used 3rd. party class visible to the area where AvroParquetRecordFormat is constructed, i.e. the factory methods? Since the constructor of AvroParquetRecordFormat is package private, it is acceptable to do it. But it is generally recommended to encapsulate such information with the class so that any changes of the 3rd. party lib on low level will not force signature changes on the higher level, e.g. avoid the Open/Closed Principle violation.

Reply:
Yes, I'd do it here since the constructor is not public and we can adjust if needed (not that I expect it since the whole model abstraction in Avro is very stable). Your point is very valid on all public accessors though.

Reply:
fair enough, I will change it.

Reply:
Unfortunately, after injecting the model, we have to hold the model as an instance variable with the type of `GenericData` which is not 'Serializable'. One solution might be override readObject(...) writeObject(...) methods and the logic will not be easier than the old one in the getDataModel(). Another option is to define a new enum for Generic as a workaround. WDYT?

Reply:
Er, tbh I forgot to check whether `GenericData` is serializable. 

While `readObject` is the most canonical solution, it's also the least developer-friendly version (there are surprisingly many devs that don't know how that works correctly). 
So from your options, I'd prefer the enum.

I'd add two more options: 3) revert to the state before my comment (duplicated logic). 4) Use `SerializableSupplier` (I'm turning it into `PublicEvolving` into another PR so that this will also work if we outsource formats).
So you have
```
AvroParquetRecordFormat(TypeInformation<E> type, SerializableSupplier<GenericData> modelSupplier) {
  this.modelSupplier = modelSupplier;
  ...
}
```
and
```
        return new AvroParquetRecordFormat<>(new AvroTypeInfo<>(typeClass), () -> GenericData.get());
```

Reply:
Thanks for the info. The 3rd option should be acceptable for now and the 4th option looks much more interesting, I will try that.